### PR TITLE
Migrate to <Section> component and extract css

### PR DESCRIPTION
**Pull request by mauriceyap**

### CODE REVIEW

File: src/components/find-out-more-text.tsx [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import styles from "./find-out-more-text.module.scss"
+
+interface FindOutMoreTextProps extends HTMLProps<HTMLDivElement> {}
+
+const FindOutMoreText: FC<FindOutMoreTextProps> = ({
+    children,
+    className,
+    ...elementProps
+}) => {
+    return (
```
Comment (mauriceyap) [inline src/components/find-out-more-text.tsx]: remove braces

File: src/components/find-out-more.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import Section from "./section"
+
+interface FindOutMoreProps extends HTMLProps<HTMLDivElement> {}
+
+const FindOutMore: FC<FindOutMoreProps> = ({
+    children,
+    className,
+    ...elementProps
+}) => {
+    return (
```
Comment (mauriceyap) [inline src/components/find-out-more.tsx]: remove braces

File: src/components/find-us.module.scss [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    &:nth-of-type(1) {
+        grid-area: buttonleft;
+    }
+
+    &:nth-of-type(2) {
+        grid-area: buttonright;
+    }
+}
+
+.map {
```
Comment (mauriceyap) [inline src/components/find-us.module.scss:117]: remove

File: src/components/find-us.tsx [hunk 4]
```diff
@@ -21,97 +24,101 @@ const FindUs = () => {
...
+                                `https://maps.google.com/maps?daddr=${latitude},${longitude}&amp;ll=`
+                            )
+                        }
+                    }}
+                >
+                    Get Directions
+                </a>
+                <a
+                    className={styles.button}
+                    href="https://citymapper.com/go/8hs1tgwo5g"
```
Comment (mauriceyap) [inline src/components/find-us.tsx]: extrac URL into static const (likewise elsewhere in this file)

File: src/components/find-us.tsx [hunk 5]
```diff
@@ -21,97 +24,101 @@ const FindUs = () => {
...
+                <InView
+                    rootMargin="0px 0px 3000px 0px"
+                    onChange={inView => {
+                        if (inView === true) {
+                            setMapHasBeenInView(true)
+                        }
+                    }}
+                >
+                    {({ ref }) => {
+                        return (
```
Comment (mauriceyap) [inline src/components/find-us.tsx]: remove braces

File: src/components/header-underlay.tsx [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+import styles from "./header-underlay.module.scss"
+
+interface HeaderUnderlayProps extends HTMLProps<HTMLDivElement> {}
+
+const HeaderUnderlay: FC<HeaderUnderlayProps> = ({
+    children,
+    className,
+    ...elementProps
+}) => {
```
Comment (mauriceyap) [inline src/components/header-underlay.tsx]: remove braces

File: src/components/basic-text-page.tsx [hunk 7]
```diff
@@ -1,20 +1,23 @@
...
 type Props = {
     html?: string
 }
 
 const BasicText: React.FC<Props> = props => {
     return (
         <>
-            <section className="header-underlay" />
-            <section>
+            <HeaderUnderlay />
```
Comment (tomduckering) [inline src/components/basic-text-page.tsx]: Nice

**Comment by tomduckering** [comment]
LGTM from a visual perspective.

**Comment by bdrobinson** [comment]
This looks like a mammoth undertaking and I like it a lot – definitely good to break this stuff out to make it more reusable.

I have a concern with the API – I have a bit of a phobia of components that expose props like `style`, `className` and let you inject any arbitrary HTML props as I think it significantly hampers maintainability of that component.

To illustrate this, imagine we have a layout component that displays a horizontal list of people's names (I'm going to use `style` rather than `className` as an example as it's a bit easier to illustrate, but the same argument applies regardless)
```ts
interface Props {
    style?: React.CSSProperties
    names: Array<string>
}
const NameList: React.FC<Props> = ({ style, names }) => {
    return (
        <div
            style={{
                display: "flex",
                flexDirection: "row",
                ...style,
            }}
        >
            {names.map(name => (
                <div>{name}</div>
            ))}
        </div>
    )
}
```

Looks okay right – you can use the `style` prop to do nice things like add padding to the component or change the font or whatever.

But now suppose some other dev else sees this component and thinks "hmm, this is exactly what I want, except I want the list to be *vertical* rather than *horizontal*. They look at the source and bam, realise all they have to do is `<NameList names={names} style={{flexDirection: "column"}} />`. It works, hurrah!

But this is actually a *big problem* as now the parent of the `<NameList />` has coupled themselves to the implementation of `NameList` itself (flexbox). Suppose in 6 months time the original dev came back and was like "hmm, we've recently dropped support for IE11 so I can refactor this from flexbox to CSS grid". They now have to manually check through every instantiation of `NameList` in the codebase (potentially dozens) to make sure that none of them passed any flexbox styles in. And if they did? Then a small refactor of one component has quickly turned into a massive PR touching a bunch of files, requiring lots of thinking and manual visual tests to make sure every instantiation of the refactored `NameList` still works.

The reason that whole mess arose is because `NameList` has waaaayyyy too big a public API, inviting all its callers to reach inside and poke at its innards in a million different ways. Instead it should do one thing and do it well. If the second dev does want to make `NameList` able to display vertically as well as horizontally, then they should make it a part of the public api (perhaps via a `direction: "vertical" | "horizontal"` prop).

The flexbox example is just the first one that sprang to mind, there's a whole bunch of other ones that would also work. Basically, a component as widely-used as this should be very aggressive in minimising its API surface area, especially with regards to CSS which is already so hard to maintain as it is.

As an aside, there are a *few* times where accepting `className` or `style` etc is appropriate – generally when you're writing a wrapper for some DOM element (ie `react-router`'s `Link` which is just a tiny wrapper around an `a`), and it's pretty unavoidable for UI libraries on npm (eg, `react-datepicker`) that need to let thousands of projects style it in their own way. But those APIs are always nasty and almost impossible for the maintainers to evolve once people have started using them. And in our case we have the luxury of knowing that our component is part of our design system so we can be very strict in limiting its scope.

Sorry this has ended up being so long – it's a bit of a bugbear of mine that the react community doesn't seem to talk about this much! I tried to find a good blog post to link to rather than write my own example but there doesn't seem to be one so I just went ahead and wrote an essay.

Please let me know what you think – definitely open to persuasion as to why the above does/doesn't apply here! But otherwise I think it would be good to go through and work out what problem is being solved each time `className` is injected and then work out if there's another way we can achieve the same.

And apologies if you'd already thought about all this and are just using the `className` API as a first pass of the refactor!

**Comment by mauriceyap** [comment]
> This looks like a mammoth undertaking and I like it a lot – definitely good to break this stuff out to make it more reusable.
> 
> I have a concern with the API – I have a bit of a phobia of components that expose props like `style`, `className` and let you inject any arbitrary HTML props as I think it significantly hampers maintainability of that component.
> 
> To illustrate this, imagine we have a layout component that displays a horizontal list of people's names (I'm going to use `style` rather than `className` as an example as it's a bit easier to illustrate, but the same argument applies regardless)
> 
> ```ts
> interface Props {
>     style?: React.CSSProperties
>     names: Array<string>
> }
> const NameList: React.FC<Props> = ({ style, names }) => {
>     return (
>         <div
>             style={{
>                 display: "flex",
>                 flexDirection: "row",
>                 ...style,
>             }}
>         >
>             {names.map(name => (
>                 <div>{name}</div>
>             ))}
>         </div>
>     )
> }
> ```
> 
> Looks okay right – you can use the `style` prop to do nice things like add padding to the component or change the font or whatever.
> 
> But now suppose some other dev else sees this component and thinks "hmm, this is exactly what I want, except I want the list to be _vertical_ rather than _horizontal_. They look at the source and bam, realise all they have to do is `<NameList names={names} style={{flexDirection: "column"}} />`. It works, hurrah!
> 
> But this is actually a _big problem_ as now the parent of the `<NameList />` has coupled themselves to the implementation of `NameList` itself (flexbox). Suppose in 6 months time the original dev came back and was like "hmm, we've recently dropped support for IE11 so I can refactor this from flexbox to CSS grid". They now have to manually check through every instantiation of `NameList` in the codebase (potentially dozens) to make sure that none of them passed any flexbox styles in. And if they did? Then a small refactor of one component has quickly turned into a massive PR touching a bunch of files, requiring lots of thinking and manual visual tests to make sure every instantiation of the refactored `NameList` still works.
> 
> The reason that whole mess arose is because `NameList` has waaaayyyy too big a public API, inviting all its callers to reach inside and poke at its innards in a million different ways. Instead it should do one thing and do it well. If the second dev does want to make `NameList` able to display vertically as well as horizontally, then they should make it a part of the public api (perhaps via a `direction: "vertical" | "horizontal"` prop).
> 
> The flexbox example is just the first one that sprang to mind, there's a whole bunch of other ones that would also work. Basically, a component as widely-used as this should be very aggressive in minimising its API surface area, especially with regards to CSS which is already so hard to maintain as it is.
> 
> As an aside, there are a _few_ times where accepting `className` or `style` etc is appropriate – generally when you're writing a wrapper for some DOM element (ie `react-router`'s `Link` which is just a tiny wrapper around an `a`), and it's pretty unavoidable for UI libraries on npm (eg, `react-datepicker`) that need to let thousands of projects style it in their own way. But those APIs are always nasty and almost impossible for the maintainers to evolve once people have started using them. And in our case we have the luxury of knowing that our component is part of our design system so we can be very strict in limiting its scope.
> 
> Sorry this has ended up being so long – it's a bit of a bugbear of mine that the react community doesn't seem to talk about this much! I tried to find a good blog post to link to rather than write my own example but there doesn't seem to be one so I just went ahead and wrote an essay.
> 
> Please let me know what you think – definitely open to persuasion as to why the above does/doesn't apply here! But otherwise I think it would be good to go through and work out what problem is being solved each time `className` is injected and then work out if there's another way we can achieve the same.
> 
> And apologies if you'd already thought about all this and are just using the `className` API as a first pass of the refactor!

Thanks  - definitely think you should make this a Medium article!

I think I largely agree with this point. The reason I decided to expose all the HTML props on the `<Section>` component here is that I initially wanted it to just be a simple wrapper around the `<section>` host element with all the CSS styles which apply to `section`. As I continued moving existing code over to use this new component, it turns out there are in a few cases where we add specific styles to `<section>` and its standard variations (dark, intro, fullBleed etc.), for example, `<Hero>` and `<HeaderUndelay>`. In these components, we add hero- and header underlay- specific styles to the `<section>`. I didn't want to just use the `<section>` host element because we would then have lost the styles that are added to it. I also didn't want to use a global CSS style like `section { ... }`.

In these cases, perhaps the solution is simply to just not extend `<Section>` and just repeat the styles required, or us a shared Sass mixin (e.g. `@include section;`).

I also want to make sure that for as many of our compoents as possible, we provide the ability to pass properties like `innerHTML`, `id`, `name` and perhaps in future, `data-test-*` through to the host element to allow for things like navigation to an anchor or testing. Perhaps the solution to this is to create an interface of `HTMLPassThroughProps` which we have our components extend and only consists of such properties.

**Pull request merged**